### PR TITLE
Add ROH, UPD and Cytoband Images to case.html

### PR DESCRIPTION
**Pull request by mikaell**
Add cytobands and other images to the dashboard page in Scout. This is a squashed commit of the work. To be easier to manage.

This PR fixes adds functionality of the Case page to display images of ideograms, UPD and ROH pictures produced by Chromograph.

**1A. How to test**:
Install this PR but without placing png:s in `public/static` folder for Scout. 
Load Demo case
Browse to Case-page

**2A. Expected outcome**:
No change. When no images are readable, the page divisions or panels will not be shown. 

Optional: Open web page inspector console and find the following message: 

    *** Not drawing ROH/Ideograms/UPDs

**1B. How to test**:
Extract ideo_upd_roh.zip under Scouts `/public/static` folder. On my installation found at:

  /miniconda3/envs/python3.6/lib/python3.6/site-packages/scout_browser-4.6.1-py3.6.egg/scout/server/blueprints/public/static/

Browse to Case-page

**2B. Expected outcome**:
The case page will display images of Roh, Upd and ideograms.
* ROH, UPD and Chromosomes Shown in Cases-page
* Make whitespace in case.html like Master (59d69b8b)
* Hide ROH-panel in case.html if the images that are suppose to be
drawn are missing.

**Review:**
- [ ] code approved by
- [ ] tests executed by
- [ ] "merge and deploy" approved by

### CODE REVIEW

**Comment by northwestwitch** [comment]
Nice, don't forget to update the branch with changes from the current master, see button 'update branch'

**Comment by mikaell** [comment]
> Nice, don't forget to update the branch with changes from the current master, see button 'update branch'

I can't see that button :( Maybe it will appear when travis run is complete or when approved.

But I pushed CHANGELOG and resolved the conflict.

-And still need a code review and tests :)

**Comment by dnil** [comment]
Thanks! Maybe someone (@moonso?) needs to give you write access for that button to appear?

**Comment by mikaell** [comment]
> Thanks! Maybe someone (@moonso?) needs to give you write access for that button to appear?

It appeared. Think Travis was blocking.

**Comment by mikaell** [comment]
Only needs code review! @dnil @northwestwitch

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 1]
```diff
@@ -280,6 +284,21 @@
   </div>
 {% endmacro %}
 
+{% macro roh_panel() %}
+<div class="panel panel-default" id="rohPanel">
+  <div data-toggle='tooltip' class="panel-heading" title="down format" >
+    Ideogram, UPD, ROH
+  </div>
+  <div class="panel-body">
+    <svg id="roh_svg" width="1900" height="850"></canvas>
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: ```<svg></canvas>```?

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 2]
```diff
@@ -54,6 +54,10 @@
           {% endif %}
         </div>
 
+        <div class="row">
+          <div class="col-xs-12 col-md-12">{{ roh_panel() }}</div>
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: :+1: for macro!

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 3]
```diff
@@ -280,6 +284,21 @@
...
+{% macro roh_panel() %}
+<div class="panel panel-default" id="rohPanel">
+  <div data-toggle='tooltip' class="panel-heading" title="down format" >
+    Ideogram, UPD, ROH
+  </div>
+  <div class="panel-body">
+    <svg id="roh_svg" width="1900" height="850"></canvas> 
+  </div>
+  <div class="panel-footer">
+    <button class="btn btn-default form-control" data-toggle="modal" data-target="#edit-synopsis">Edit</button>
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: Edit synopsis? Nah, that should only be on the synopsis_panel!

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 4]
```diff
@@ -930,4 +949,327 @@ <h4 class="modal-title">Edit synopsis</h4>
...
+                        'upd_chr4.png', 'upd_chr5.png', 'upd_chr6.png', 
+                        'upd_chr7.png', 'upd_chr8.png', 'upd_chr9.png',
+                        'upd_chr10.png', 'upd_chr11.png', 'upd_chr12.png',
+                        'upd_chr13.png', 'upd_chr14.png', 'upd_chr15.png',
+                        'upd_chr16.png', 'upd_chr17.png', 'upd_chr18.png',
+                        'upd_chr19.png', 'upd_chr20.png', 'upd_chr21.png',
+                        'upd_chr22.png', 'upd_chrX.png', 'upd_chrY.png'];
+
+
+    // TODO: Handle trisomy
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: :smile:

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 5]
```diff
@@ -930,4 +949,327 @@ <h4 class="modal-title">Edit synopsis</h4>
...
+                         '19', '20', '21', '22', 'X',' Y'];
+
+
+
+
+    const NUMBER_OF_COLUMNS = 3        
+    const OFFSET_X = 60;
+    const OFFSET_Y = 30;
+
+    // skära ut kromosom
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: We try to stick to English for comments!

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 6]
```diff
@@ -930,4 +949,327 @@ <h4 class="modal-title">Edit synopsis</h4>
...
+    // "M 300 150 L 300 500 L 500 500 A 50 50 0 1 1 500 150 L 300 150"
+
+    
+
+    // Draw ROH call pictures -coverage- and UPOD pictures -color coded
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: UPOD -> UPD

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 7]
```diff
@@ -930,4 +949,327 @@ <h4 class="modal-title">Edit synopsis</h4>
...
+    // genome regions- onto the dashboard.
+    //
+    //
+    function draw_tracks(){
+        const CYT_HEIGHT = 50 ;
+        const CYT_WIDTH = 500 ;
+        const ROH_IMGPATH = '/public/static/roh_images/'
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: Ok, here we need to work together. For all files that do change, the scout load case function should read roh file (and image) paths from the case config file, and pass the path here.

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 8]
```diff
@@ -930,4 +949,327 @@ <h4 class="modal-title">Edit synopsis</h4>
...
+        console.log("draw roh") 
+        var roh_svg = document.getElementById("roh_svg");
+
+        var roh_imgObj = new Image();
+        
+        var upd_imgPath = '/public/static/upd_images/'
+        var upd_imgObj = new Image();
+        // Loop through to add ROH and UPD images
+
+        var chrom1 = {name: '1', length: 482, cent_start: 245, cent_length: 13 }
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: So, I guess these could stay state until we get a hg38 - but given time we should prepare for at least two different versions

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 9]
```diff
@@ -930,4 +949,327 @@ <h4 class="modal-title">Edit synopsis</h4>
...
+        const CYT_WIDTH = 500 ;
+        const ROH_IMGPATH = '/public/static/roh_images/'
+        const IDEO_IMGPATH = '/public/static/chr_images/'
+        
+        console.log("draw roh") 
+        var roh_svg = document.getElementById("roh_svg");
+
+        var roh_imgObj = new Image();
+        
+        var upd_imgPath = '/public/static/upd_images/'
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: ..and same here!

File: scout/server/blueprints/cases/templates/cases/case.html [hunk 10]
```diff
@@ -930,4 +949,327 @@ <h4 class="modal-title">Edit synopsis</h4>
...
+   }
+  </script>
+
+
+  <script>
+    // Start when window is loaded!
+    window.onload=function(){
+        console.log("Window is loaded!")
+        console.log("Draw all cromosome data tracks to svg image area")
+        var testFile = "/public/static/chr_images/cytoband.txt.chr1.png"
```
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/case.html]: And here as well - should be tested with one of the changeable files.

File: scout/server/blueprints/cases/templates/cases/case.html (hunk 7 above)
Comment (mikaell) [reply scout/server/blueprints/cases/templates/cases/case.html]: Yes! I'm thinking two config parameters:

    chromograph_image_location: /some/path
    chromograph_image_destination: /some/other/path

-But some other name?

**Comment by dnil** [comment]
Since I haven’t been fighting with the particular issue necessitating file relocation I still find it unnecessary with more than one (original) path? It works for bam-files, and I can’t quite wrap my head around why it shouldn’t work to make the files accessible in the first place. Perhaps we can start with one path, then change it if we just can’t make it fly through cg/housekeeper? 

// Daniel

File: scout/server/blueprints/cases/templates/cases/case.html (hunk 6 above)
Comment (mikaell) [reply scout/server/blueprints/cases/templates/cases/case.html]: Fixed!

File: scout/server/blueprints/cases/templates/cases/case.html (hunk 5 above)
Comment (mikaell) [reply scout/server/blueprints/cases/templates/cases/case.html]: Fixed! (0eb3edc)

File: scout/server/blueprints/cases/templates/cases/case.html (hunk 3 above)
Comment (mikaell) [reply scout/server/blueprints/cases/templates/cases/case.html]: Fixed! (0eb3edc)

File: scout/server/blueprints/cases/templates/cases/case.html (hunk 1 above)
Comment (mikaell) [reply scout/server/blueprints/cases/templates/cases/case.html]: Fixed! (0eb3edc)

File: scout/server/blueprints/cases/templates/cases/case.html (hunk 10 above)
Comment (dnil) [reply scout/server/blueprints/cases/templates/cases/case.html]: (at some point.. :wink:

File: scout/server/blueprints/cases/templates/cases/case.html (hunk 7 above)
Comment (mikaell) [reply scout/server/blueprints/cases/templates/cases/case.html]: In 4dbe309e2 I've added configurable path to image dir. Though some stuff is still static. The code expects 

1. Three separate subdirs, one for each file type
     <config_path/>
        chr_images/
         roh_images/ 
         upd_images/

2. Fixed filenames
cytoBand.txt.chr1.png
upd_chr1.png
roh_chr1.png

-I think keeping strict directory names is good in order not to mess it up. However, static file names is not.

### Solution A

.*([1-22]|X|Y)\.png

(regex might be slightly off, intension is to match every png-file which ends with 1 through 22 or X or Y)


### Solution B

Configure prefixes -still will do some sort of reqexp match..


What do you think?

**Comment by mikaell** [comment]
I quickly implemented 

### Solution B
Configure prefixes -still will do some sort of reqexp match..

In <case>.config.yaml you specify
```
chromograph_prefixes:
    'roh': "roh_chr"
    'upd': "upd_chr"
    'chr': "cytoband.txt.chr"
```

In javascript I then loop through adding suffixes 1 though 22 plus X and Y. This makes it more custom for the user and easier on the user.

What do you think?

**Comment by dnil** [comment]
Good. I don't think you have to worry too much. Fixed filenames are a bit crude, but as long as we keep Chromograph and Scout synced that is ok. The important thing to realise is just that one or more of the tracks could be missing entirely. E.g. singletons would have no UPD track.

**Comment by mikaell** [comment]
I'm working on finally closing this pull request. Have added:
* Check if config params exists, otherwise don't show images
* Adapt number of columns after window width (this is statical, done on load, someday it should be dynamic)
* Show correct sex-chromosomes

But I saw that in Bootstrap 4, when displaying Chromograph images, the panels get "pushed down". See image. I'm trying to fix this now. 

Other smaller issues, such as nicer css (borders are a bit tight), can be done in separate PR:s. I really want to close this one :) 


Screenshot 2019-10-19 at 15 54 29

**Comment by dnil** [comment]
Let me know when its time to test / review, or if you want more input! Good enough is good enough - it will not show until chronograph is live in production, so there is still time for the fine print later. Wrapping this up and moving to wig/bed load to IGV is more important!

**Comment by moonso** [comment]
I agree with @dnil , try to merge this now if it has some functionality in place and then we can iterate later. PRs just get harder and harder to merge with age...

**Comment by mikaell** [comment]
Merged everything and the history got so badly messed up… So I created a new branch and a new PR: #1410

Exactly the same functionality waiting to be reviewed :)

**Comment by dnil** [comment]
Thanks - I'll close this one then - looking at #1410 next!

**Pull request closed without merging**